src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = [];
+  let clone = Object.assign({}, state);
+  for (let i = 0; i < actions.length; i++) {
+    const { type, extraData, keysToRemove } = actions[i];
+
+    if (i > 0) {
+      clone = Object.assign({}, newState[i - 1]);
+    }
+
+
+    if (type === 'addProperties') {
+      newState.push({ ...clone, ...extraData });
+    }
+
+    if (type === 'removeProperties') {
+      for (const key in clone) {
+        for (let l = 0; l < keysToRemove.length; l++) {
+          if (key === keysToRemove[l]) {
+            delete clone[key];
+          }
+        }
+      }
+      newState.push(clone);
```

Comment:
Use a more efficient way to remove properties. Instead of nested loops, you can check if 'keysToRemove' includes the key and delete it if true. Then, push the modified clone to 'newState' after processing the action outside of the if statement.